src/System.Net.Security/src/System/Net/Unix/SSPISecureChannelType.cs

```diff
@@ -0,0 +1,290 @@
+                }
+
+                context.DangerousAddRef(ref gotContextReference);
+           
+                IntPtr inputPtr = IntPtr.Zero, outputPtr = IntPtr.Zero;
+                int outputSize;
+                bool done;
+
+                if (null == inputBuffer)
+                {
+                    done = Interop.OpenSsl.DoSslHandshake(context.DangerousGetHandle(), inputPtr, 0, out outputPtr, out outputSize);
+                }
+                else
+                {
+                    inputHandle = GCHandle.Alloc(inputBuffer.token, GCHandleType.Pinned);
+                    inputPtr = Marshal.UnsafeAddrOfPinnedArrayElement(inputBuffer.token, inputBuffer.offset);
+                    done = Interop.OpenSsl.DoSslHandshake(context.DangerousGetHandle(), inputPtr, inputBuffer.size, out outputPtr, out outputSize);
+                }           
+
+                outputBuffer.size = outputSize;
+                outputBuffer.offset = 0;
+                if (outputSize > 0)
+                {
+                    outputBuffer.token = new byte[outputBuffer.size];
+                    Marshal.Copy(outputPtr, outputBuffer.token, 0, outputBuffer.size);
+                    Marshal.FreeHGlobal(outputPtr);
+                }
+                else
+                {
+                    outputBuffer.token = null;
+                }      
+
+                return done ? SecurityStatus.OK : SecurityStatus.ContinueNeeded;
+            }
+            catch
+            {
+                return SecurityStatus.InternalError;
+            }
+            finally
+            {
+                if (inputHandle.IsAllocated)
+                {
+                    inputHandle.Free();
+                }
+                if (gotContextReference)
+                {
+                    context.DangerousRelease();
+                }
+                if (gotCredReference)
+                {
+                    credential.DangerousRelease();
+                }
+            }
+        }
+
+
+        private SecurityStatus EncryptDecryptHelper(SafeDeleteContext securityContext, byte[] buffer, int offset, int size, int headerSize, int trailerSize, bool encrypt, out int resultSize)
+        {
+            bool gotReference = false;
+            GCHandle inputHandle = new GCHandle();
+            resultSize = 0;
+            try
+            {
+                securityContext.DangerousAddRef(ref gotReference);
+                inputHandle = GCHandle.Alloc(buffer, GCHandleType.Pinned);
+                IntPtr inputPtr = Marshal.UnsafeAddrOfPinnedArrayElement(buffer, 0);
+                if (encrypt)
+                {
+                    resultSize = Interop.OpenSsl.Encrypt(securityContext.DangerousGetHandle(), inputPtr, offset, size, buffer.Length);
+                }
+                else
+                {
+                    resultSize = Interop.OpenSsl.Decrypt(securityContext.DangerousGetHandle(), inputPtr, size);
```

Comment:
I'll stop asking about these for each of the methods, but it'd be great if we could pass the SafeHandle down as close to the native calls as possible.